<a href="https://colab.research.google.com/github/juan-scholarium/bandungpy/blob/main/Pengenalan_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INTRODUCTION TO PYTHON


Referensi:

1. [The Python Workshop](https://www.google.co.id/books/edition/The_Python_Workshop/zV3HywEACAAJ?hl=en) ==> Referensi utama
2.[Learn Python Programming](https://www.google.co.id/books/edition/Learn_Python_Programming/NsNiDwAAQBAJ?hl=en&gbpv=0)
3. [Data Structure and Algorithms in Python](https://www.google.co.id/books/edition/Data_Structures_and_Algorithms_in_Python/Mr59DwAAQBAJ?hl=en&kptab=overview)



# What is Algorithms?

* Algoritma adalah [prosedur sistematis untuk memecahkan masalah matematis dalam langkah-langkah terbatas](https://kbbi.kemdikbud.go.id/entri/algoritma).
* Dalam kesempatan kali ini kita akan berkenalan terlebih dahulu dengan Python yang menjadi bahasa untuk mengekspresikan Algoritma.

# Why Python?


* Mudah untuk dibaca
* Memiliki *library* yang banyak
* Menurut Mark Lutz, program Python umumnya memiliki ukuran 1/5 sampai 1/3 dibandingkan dengan Java atau C++ (Learning Python, 5th Edition, O'Reilly Media)
* Salah satu bahasa pemrograman yang paling bertumbuh penggunanya (lihat referensi trend Python)

**Referensi trend Python**
* [PYPL PopularitY of Programming Language](https://pypl.github.io/PYPL.html)
* [Google Trends](https://trends.google.co.id/trends/explore?date=all&q=%2Fm%2F05z1_,%2Fm%2F02p97)

# Where can I get Python?


* [Official Web Site Python](https://www.python.org/)
* [Anaconda](https://www.anaconda.com/products/individual)
* [Google Colab](https://colab.research.google.com/) ==> ini yang akan kita gunakan
* [Microsoft Store](https://www.microsoft.com/en-us/p/python-37/9nj46sx7x90p#activetab=pivot:overviewtab) (khusus Windows)

# Our Scope 


* Data Types - integer, float, string, boolean
* Vital Python - Math, Strings, Conditionals, Loops, and Function
* Python Structure - List, Dictionary, Tuple, Set


---



# Data Types

* Tipe data ini akan berpengaruh terhadap operasi yang diterapkan terhadap data. Misal, tipe data `int` untuk perhitungan, `boolean` untuk mengetahui kondisi suatu variabel.

* Pada Python, tipe data tidak harus didefinisikan terlebih dahulu.

* A script is simply a collection of statements

In [189]:
# tipe data integer
type(1)

int

In [190]:
# tipe data float
type(1.0)

float

In [191]:
# tipe data string
type("algoritma")

str

In [192]:
# tipe data boolean
type(False)

bool

# Vital Python - Math, Strings, Conditionals, and Loops + Function

## Numbers: Operations and Variables


A number can be either an integer or a decimal (floating point).

### Operasi matematika

In [193]:
# Penjumlahan
1 + 2

3

In [194]:
# Pengurangan
3 - 1

2

In [195]:
# Pembagian (float)
4 / 2

2.0

In [196]:
# Pembagian (integer)
7 // 3

2

In [197]:
# Perkalian
2 * 2

4

In [198]:
# Pangkat
2 ** 3

8

In [199]:
# Modulus
3 % 2

1

**Take away**

Perhitungan matematika pada Python mengikuti ordo perhitungan matematika pada umumnya (parentheses first, exponentiation second, multiplication/division third, and addition/subtraction fourth)

In [200]:
# Contoh ordo perhitungan
(1 + 2) * 3 / 4 + 1

3.25

Urutannya:


1.   1 + 2 = 3
2.   3 * 3 = 9
3.   9 / 4 = 2.25
4.   2.25 + 1 = 3.25



### Variabel

* A variable is a name associated with a piece of data 
* Variables allow you to store and manipulate data in your programs
* Think of a variable as a mailbox which holds a specific piece of information
* It is vitally important to distinguish between the name of the variable and the value of the variable.


**Programming Tips**
* If a variable is initialized as a number, it should always be used as an number.
* Choose meaningful variable names

In [201]:
# Nilai dapat disimpan dalam sebuah variabel.
# Variabel ini dapat digunakan berulang kali
# Variabel bersifat case sensitif

x = 3
y = 4

luas = x * y # hasil perhitungan disimpan dalam variabel "luas"
print(luas) # tampilkan nilai dari "luas"

12


In [202]:
# Variabel dapat diubah dengan nilai lain
x = 3
y = 4

# ganti variabel x
x = 5

luas = x * y
print(luas)

20


**Take away**

Nama variabel bisa apa pun selain *reserved words* dan tidak diawali dengan angka, terdapat tanda baca, spasi

In [203]:
# Reserved words
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


In [204]:
# contoh variabel yang diawali dengan angka
1angka = 1

SyntaxError: ignored

## Strings


A string is a sequence of letters or numbers enclosed in single or double quotes

### Concatenation, Methods, and input()

In [205]:
'this is a string type data' # tipe data string diapit oleh tanda " atau '

'this is a string type data'

In [206]:
type ("this is a string type data")

str

In [207]:
# string dalam variabel
kalimat = "belajar Algoritma dalam bahasa Python"
print(kalimat)

belajar Algoritma dalam bahasa Python


In [208]:
# Concatenation
print("Juan " + kalimat)

Juan belajar Algoritma dalam bahasa Python


In [209]:
# Panjang dari sebuah data string
len("Juan ")

5

In [210]:
# String Methods
name = "Juan"
name.lower() # eksplorasi lainnya

'juan'

In [211]:
# Input()
name = input()
print("Halo ", name.capitalize())

juan
Halo  Juan


### String Indexing & Slicing

Indexing: mengambil spesifik (satu) elemen \\
Slicing: mengambil sebagian (kumpulan) elemen

In [212]:
kata = "Algoritma" # huruf "A" berada di index 0, huruf "a" berada di index 8 karena index dimulai dari 0
len(kata)

9

In [213]:
# mengambil huruf di index 0
kata[1]

'l'

In [214]:
kata[-2] # index negatif berarti kita mengambil index dari urutan belakang

'm'

In [215]:
# mengambil sebagian elemen / slicing
kata[0:3]

'Alg'

**Take away**

Indexing dan slicing dapat diaplikasikan pada Array atau List, DataFrame

## Booleans, Comparison Operator, and Conditionals

Tipe data booleans sangat erat kaitannya dengan operator pembanding dan conditionals (`if`)



### Booleans

Ingat tabel logika?

![Tabel Logika](https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Finstrumentationtools.com%2Fwp-content%2Fuploads%2F2017%2F07%2Finstrumentationtools.com_digital-logic-gates-truthtables.png&f=1&nofb=1)

### Operator Pembanding

In [216]:
# Operator pembanding lebih kecil dari (<)
1 < 2

True

In [217]:
# Operator pembanding lebih kecil sama dengan (<=)
2 <= 2

True

In [218]:
# Operator pembanding lebih besar dari (>)
1 > 2

False

In [219]:
# Operator pembanding lebih besar sama dengan (>=)
1 >= 2

False

In [220]:
# Operator pembanding sama dengan (==)
# Hint: operator == berbeda dengan =
1 == 2

False

In [221]:
# Operator pembanding tidak sama dengan (!=)
1 != 2

True

### Conditionals


Contoh `conditionals` dalam kehidupan sehari-hari


* **Jika** umur belum 17 tahun, **maka** belum boleh mengemudi.
* **Jika** sudah vaksin, **maka** boleh makan di tempat


Use an if statement whenever you come to a “fork” in the program

In [222]:
age = 1

if age < 17: #Jika True
    print("Belum boleh mengemudi") # indentasi sangat penting pada Python
else:
    print("Boleh mengemudi")

Belum boleh mengemudi


In [223]:
vaksin = "Belum"

if vaksin == "Sudah":
    print("Anda boleh makan di tempat")
else:
    print("Anda tidak boleh makan di tempat")

Anda tidak boleh makan di tempat


In [224]:
# Condisional bisa bertingkat
vaksin = "aaa"

if vaksin == "Sudah": #Jika True
    print("Anda boleh makan di tempat")
elif vaksin == "Belum":
    print("Anda tidak boleh makan di tempat")
else:
    print("Tolong sebutkan yang benar, Anda sudah divaksin atau belum?")

Tolong sebutkan yang benar, Anda sudah divaksin atau belum?


## Loops


Contoh `loops` dalam kehidupan sehari-hari:

* Tulis "Aku belajar Python" sebanyak 100 kali.


Terdapat tiga konsep utama dalam looping


1.   Permulaan looping
2.   Akhir looping
3.   Increment selama proses looping


### `while` loop


The while loop is used to execute a block of code while a certain condition is true

In [225]:
kalimat = "Aku belajar Python"
i = 0

while i < 5:
    print(kalimat)
    i += 1 # sama dengan i = i + 1, istilahnya aggregate assignments

Aku belajar Python
Aku belajar Python
Aku belajar Python
Aku belajar Python
Aku belajar Python


### `for` loop


* The for loop is used when there is a need to have a counter of some kind
* The counter is initialized before the loop starts, tested after each iteration to see if it is below a target value, and finally updated at the end of the loop

In [226]:
for i in range(3):
    print(kalimat)
    i += 1

Aku belajar Python
Aku belajar Python
Aku belajar Python


## Function


* Functions are a collection of statement that performs a specified task
* Functions are used whenever it is necessary to repeat an operation
* Functions have inputs and outputs
* The inputs are passed into the function and are known as arguments or parameters
* Think of a function as a “black box” which performs an operation.

In [227]:
# contoh function penjumlahan
def eq_add(x, y):
    return(x + y)

print("Jumlah = ", add_value(2, 3))

Jumlah =  5


In [228]:
# contoh function menghitung luas dan volume balok
def eq_luas(p, l, t):
    luas = 2 * (p * l + p * t + l * t)
    return(luas)

def eq_volume(p, l, t):
    volume = p * l * t
    return(volume)

In [229]:
# ukuran balok yang kita miliki
p = 2
l = 3
t = 4

In [230]:
# hitung luas menggunakan fungsi yang sudah dibuat
print("Luas balok = ", eq_luas(p, l, t))

# hitung volume menggunakan fungsi yang sudah dibuat
print("Volume balok = ", eq_volume(p, l, t))

Luas balok =  52
Volume balok =  24


# Python Structures


* Struktur data `list, dictionary, tuple, set` digunakan untuk menyimpan data yang lebih kompleks (dibandingkan dengan variabel).
* Struktur data ini mendefinisikan relasi antara data dan operasi yang dapat dilakukan pada data tersebut.

## Lists

* List adalah tipe kontainer pada Python yang digunakan untuk menyimpan kumpulan data dalam waktu yang bersamaan.
* List pada Python seringkali dibandingkan dengan array dalam bahasa pemrograman lain.
* List pada Python ditandai dengan `[]`
* List dapat diakses dengan index (mulai dari 0)

In [231]:
# membuat list
daftar_belanja = ['apel', 'jeruk', 'pisang', 'anggur']
print(daftar_belanja)

['apel', 'jeruk', 'pisang', 'anggur']


In [232]:
# mengakses list
daftar_belanja[1]

'jeruk'

In [233]:
# menambah item ke dalam list
daftar_belanja.append('semangka')
daftar_belanja

['apel', 'jeruk', 'pisang', 'anggur', 'semangka']

In [234]:
# insert item ke dalam list
daftar_belanja.insert(1, 'melon')
daftar_belanja

['apel', 'melon', 'jeruk', 'pisang', 'anggur', 'semangka']

In [235]:
# mengganti item di dalam list
daftar_belanja[1] = 'ikan mas'
daftar_belanja

['apel', 'ikan mas', 'jeruk', 'pisang', 'anggur', 'semangka']

In [236]:
# banyaknya item dalam list
len(daftar_belanja)

6

In [237]:
# iterasi pada list
for item in daftar_belanja:
    print("buah " + item)

buah apel
buah ikan mas
buah jeruk
buah pisang
buah anggur
buah semangka


In [238]:
# nested list
x = [[1, 2, 3, 3],[4, 5, 6, 7], [11, 12, 13, 14]] # row, column

for i in x:
    print(i)

[1, 2, 3, 3]
[4, 5, 6, 7]
[11, 12, 13, 14]


In [239]:
# mengakses nested list
print(x[0][1])

2


## Dictionary Keys and Values


* A Python dictionary is an unordered collection.
* Dictionaries are written with curly brackets, and they have keys and values.
* Keys must be unique — no duplicate keys are allowed.
* Keys must be immutable — they can be a string, a number, or a tuple.

In [240]:
employee = {'name': "Jack Nelson",
            'age': 32,
            'department': "sales"}

In [241]:
employee['name']

'Jack Nelson'

In [242]:
# membuat dictionary dari blank
movie = {}
movie['title'] = "The Godfather"
movie['director'] = "Francis Ford Coppola"
movie['year'] = 1972
movie['rating'] = 9.2

In [243]:
# menambah key-value ke dalam dictionary movie
# list dalam dictionary
# dictionary dalam dictionary
movie['actors'] = ['Marlon Brando', 'Al Pacino', 'James Caan']
movie['other_details'] = {'runtime': 175,
                          'language': 'English'}

In [244]:
print(movie)

{'title': 'The Godfather', 'director': 'Francis Ford Coppola', 'year': 1972, 'rating': 9.2, 'actors': ['Marlon Brando', 'Al Pacino', 'James Caan'], 'other_details': {'runtime': 175, 'language': 'English'}}


## Tuples


* A tuple object is similar to a list, but it cannot be changed. 
* Tuples are immutable sequences, which means their values cannot be changed after initialization

In [245]:
# nama-nama hari dalam struktur list
weekdays_list = ['Monday', 'Tuesday', 'Wednesday','Thursday','Friday','Saturday', 'Sunday']

# nama-nama hari dalam struktur tuple
weekdays_tuple = ('Monday', 'Tuesday', 'Wednesday','Thursday','Friday','Saturday', 'Sunday')

In [246]:
# indexing tuple
print(weekdays_tuple[1])

Tuesday


In [247]:
# menambahkan hari
weekdays_tuple = weekdays_tuple + ('new_day', 'old_day')
print(weekdays_tuple)

('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday', 'new_day', 'old_day')


In [248]:
# menambahkan satu value ke dalam tuple
weekdays_tuple = weekdays_tuple + ('great_day',)
print(weekdays_tuple)

('Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday', 'new_day', 'old_day', 'great_day')


In [249]:
weekdays_tuple[0] = 'same_day'

TypeError: ignored

## Set

* They are unordered (original order are not preserved) collections of unique and immutable objects that support operations mimicking mathematical set theory.
* As sets do not allow multiple occurrences of the same element, they can be used to effectively prevent duplicate values.
* A set is a collection of objects (called members or elements)

![Set](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.gmatsyllabus.com%2Fimages%2Fstories%2Fquant%2Fvenn-diagrams.png&f=1&nofb=1)

In [250]:
# contoh set
s1 = set([1,2,3,4,5,6])
print(s1)

s2 = set([1,2,2,3,4,4,5,6,6])
print(s2)

s3 = set([3,4,5,6,6,6,1,1,2])
print(s3)

s4 = {"apple", "orange", "banana", "apple"}
print(s4)

{1, 2, 3, 4, 5, 6}
{1, 2, 3, 4, 5, 6}
{1, 2, 3, 4, 5, 6}
{'banana', 'orange', 'apple'}


In [251]:
# menambah item ke dalam set
s4.add('pineapple')
print(s4)

{'pineapple', 'banana', 'orange', 'apple'}


In [252]:
# contoh lainnya
s5 = {1,2,3,4}
s6 = {3,4,5,6}

s7 = {1,2,3}
s8 = {1,2,3,4,5}

In [253]:
# union
print("union :", s5.union(s6))

# intersection
print("intersection :", s5.intersection(s6))

# difference
print("difference :", s5.difference(s6))

# subset
print("subset :", s5.issubset(s6))
print("subset :", s7.issubset(s8))

# superset
print("superset :", s8.issuperset(s7))

union : {1, 2, 3, 4, 5, 6}
intersection : {3, 4}
difference : {1, 2}
subset : False
subset : True
superset : True
